KIPs/kip-149.md

```diff
@@ -0,0 +1,292 @@
+
+```go
+if chain.Config().IsKIP149ForkBlock(header.Number) {
+	// Inject the bytecode and states for the Registry
+	err := registry.InstallRegistry(state, chain.Config().RegistryInit)
+	if err != nil {
+		logger.Error("failed to set the registry contract code", "err", err)
+	} else {
+		logger.Info("successfully set the registry contract code", "block", header.Number.Uint64())
+	}
+}
+```
+
+#### Resolver
+
+A Klaytn node will have a resolver to read the active addresses of system contracts from the Registry. But the system contracts deployed by state injection (e.g., `KIP113`) will be directly read from `chain.Config().RegistryInit` at the `Kip149CompatibleBlock` since the Registry will be deployed in the `engine.Finalize` function, which is the last part of the block generation. In other words, the resolver will be used starting in the `Kip149CompatibleBlock + 1`.
+
+```go
+// Note that it will be used starting in the next of kip-149 fork block
+func ReadRegistryActiveAddr(backend bind.ContractCaller, name string, num *big.Int) (common.Address, error) {
+	code, err := backend.CodeAt(context.Background(), RegistryAddr, num)
+	if err != nil {
+		return common.Address{}, err
+	}
+	if code == nil {
+		return common.Address{}, ErrRegistryNotInstalled
+	}
+
+	caller, err := contracts.NewRegistryCaller(RegistryAddr, backend)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	opts := &bind.CallOpts{BlockNumber: num}
+	return caller.GetActiveAddr(opts, name)
+}
+```
+
+## Rationale
+
+### Bytecode Injection for Registry Deployment
+
+In the future, all system contracts will be registered in the Registry, and a Klaytn node will read the active addresses of system contracts from the Registry. Not only for a Klaytn node but also for other ecosystem participants who will use the registry to read the system contracts they need, meaning the registry should be registered at an easily accessible reserved address.
+
+### Delegating Ownership of Registry and System Contracts to Governance
+
+The Registry holds all system contracts for Klaytn, affecting the protocol directly. This means its registration and deprecation process must be very careful and not centralized. The same goes for the system contract. As mentioned in [specifictaion](#specification), making the EOA the owner of system contracts can cause centralization and potential private key loss issues. By delegating ownership of the Registry and system contracts to Governance, all registry and system contract changes will only be applied after full discussion and consensus of the GCs and Validators. Validators have the right to accept or reject the proposal depending on whether or not the hard fork proceeds.
```

Comment:
```suggestion
The Registry holds all system contracts for Klaytn, affecting the protocol directly. This means its registration and deprecation process must be very careful and not centralized. The same goes for the system contract. As mentioned in [specification](#specification), making the EOA the owner of system contracts can cause centralization and potential private key loss issues. By delegating ownership of the Registry and system contracts to Governance, all registry and system contract changes will only be applied after full discussion and consensus of the GCs and Validators. Validators have the right to accept or reject the proposal depending on whether or not the hard fork proceeds.
```

Reply:
Applied in https://github.com/klaytn/kips/pull/150/commits/effc933e4b643391da3359ce1aec6bfb4c9e063a